Enhance tfact_problem_events with Submission Details

quazi-h opened the pull request:
### What are the relevant tickets?
https://github.com/mitodl/hq/issues/7195

### Description (What does it do?)
Combine data from tracking logs, studentmodule, and studentmodulehistoryextended in order to enhance the tfact_problem_events data used for reporting student activity.

**Affected tables:**
tfact_problem_events

### How can this be tested?
```
dbt run --select tfact_studentmodule_problems
dbt build --select tfact_problem_events
```

rachellougee commented on src/ol_dbt/models/dimensional/tfact_problem_events.sql:
```diff
@@ -101,6 +101,191 @@ with mitxonline_problem_events as (
     select * from {{ ref('dim_platform') }}
 )
 
+, content as (
```
There should be a dedupe query in this file, as simply doing a UNION ALL on data from studentmodule would introduce duplicates.

quazi-h commented on src/ol_dbt/models/dimensional/_dim__models.yml:
```diff
@@ -465,8 +465,58 @@ models:
+    description: int, incremental attempt number generated when parsing answer submissions
+  - name: success
+    description: str, indicating whether an answer is correct or incorrect
+  - name: grade
+    description: float, floating point value indicating the total unweighted grade
+      for this problem that the learner has scored. e.g. how many responses they got
+      right within the problem. This data is also available in courseactivity_problemcheck
```
Updating the description to "number"

quazi-h commented on src/ol_dbt/models/dimensional/_dim__models.yml:
```diff
@@ -465,8 +465,58 @@ models:
+    description: str, unique identifier for a problem in the studentmodule state data
+  tests:
+  - dbt_expectations.expect_compound_columns_to_be_unique:
+      column_list: ["user_id", "courserun_readable_id", "problem_id", "attempt"]
```
Updated to match the tfact_problem_events composite unique key:
tfact_problem_events:
      column_list: ["openedx_user_id", "courserun_readable_id", "problem_block_fk",
        "event_timestamp"]
tfact_student_mdoule_problems:
      column_list: ["openedx_user_id", "courserun_readable_id", "problem_block_id", "event_timestamp"]

quazi-h commented:
I reworked the logic I was using to explode each student module record because I had been trying to extract each individual problem per problem block but is now how we were treating the data from the tracking logs. I believe we should be handling each problem block as an individual unit since that is how the data is structured under dim_content. I also removed the conditional logic I added to handle edxorg since that source does not have historyextended data and I removed edxorg processing completely from this model.

Rachel suggested that we might as well just use student module data instead of mixing it in with the tracking log data and then spending more time on deduping the records between the two, especially since we had intended to switch over to sourcing this data from the student modules in the first place.

Currently, tfact_problem_events is pulling the "problem_check" and "show_answer" events from the tracking logs. I am testing the overlap between the two sources of problem data by running tfact_problem_events and only pulling "problem_check" events and then comparing what is pulled against what's in tfact_studentmodule_problems. (Excluding "show_anwer" events because we aren't sure of actual usage).

Currently, both models are building successfully, and my counts confirm that the records in tfact_problem_events are distinct on openedx_user_id, courserun_readable_id, problem_block_fk, attempt, so this PR should be good to go. We may want to do the work to remove the tracking log data in a separate ticket so as not to overcomplicate this PR.

rachellougee reviewed (commented):
There are 5676408 duplicates when I ran dbt build --select dbt build --select tfact_studentmodule_problems. I think its related to the join with dim_content. Can you address that?

rachellougee reviewed (approved):
LGMT now. Thanks for addressing the feeback
